contracts/metatx/ERC2771Forwarder.sol

```diff
@@ -0,0 +1,240 @@
+ * * `data`: Encoded `msg.data` to send with the requested call.
+ */
+contract ERC2771Forwarder is EIP712, Nonces {
+    using ECDSA for bytes32;
+
+    struct ForwardRequest {
+        address from;
+        address to;
+        uint256 value;
+        uint256 gas;
+        uint256 nonce;
+        uint48 deadline;
+        bytes data;
+    }
+
+    bytes32 private constant _FORWARD_REQUEST_TYPEHASH =
+        keccak256(
+            "ForwardRequest(address from,address to,uint256 value,uint256 gas,uint256 nonce,uint48 deadline,bytes data)"
+        );
+
+    /**
+     * @dev Emitted when a `ForwardRequest` is executed.
+     */
+    event ExecutedForwardRequest(address indexed signer, uint256 nonce, bool success, bytes returndata);
+
+    /**
+     * @dev The request `from` doesn't match with the recovered `signer`.
+     */
+    error ERC2771ForwarderInvalidSigner(address signer, address from);
+
+    /**
+     * @dev The requested `value` doesn't match with the available `msgValue`, leaving ETH stuck in the contract.
+     */
+    error ERC2771ForwarderMismatchedValue(uint256 value, uint256 msgValue);
+
+    /**
+     * @dev The list of requests length doesn't match with the list of signatures length.
+     */
+    error ERC2771ForwarderInvalidBatchLength(uint256 requestsLength, uint256 signaturesLength);
+
+    /**
+     * @dev The request `deadline` has expired.
+     */
+    error ERC2771ForwarderExpiredRequest(uint256 deadline);
+
+    /**
+     * @dev See {EIP712-constructor}.
+     */
+    constructor(string memory name) EIP712(name, "1") {}
+
+    /**
+     * @dev Returns `true` if a request is valid for a provided `signature` at the current block.
+     *
+     * NOTE: A request with an invalid nonce will return false here but won't revert to prevent revert
+     * when a batch includes a request already executed by another relay.
+     */
+    function verify(ForwardRequest calldata request, bytes calldata signature) public view virtual returns (bool) {
+        (bool alive, bool signerMatch, bool nonceMatch, ) = _validate(request, signature);
+        return alive && signerMatch && nonceMatch;
+    }
+
+    /**
+     * @dev Executes a `request` on behalf of `signature`'s signer guaranteeing that the forwarded call
+     * will receive the requested gas and no ETH is stuck in the contract.
+     */
+    function execute(
+        ForwardRequest calldata request,
+        bytes calldata signature
+    ) public payable virtual returns (bool, bytes memory) {
+        if (msg.value != request.value) {
+            revert ERC2771ForwarderMismatchedValue(request.value, msg.value);
+        }
+
+        (bool success, bytes memory returndata) = _execute(request, signature);
+
+        return (success, returndata);
+    }
+
+    /**
+     * @dev Batch version of {execute}.
+     */
+    function executeBatch(ForwardRequest[] calldata requests, bytes[] calldata signatures) public payable virtual {
```

Comment:
Is it necessary to submit this as two separate arrays? Could we do a single array with a struct inside with request + signature?

Reply:
Given that now `ForwardRequest` is not the same as `ForwardRequestData`, we agreed to include the signature along with the request data.